### ISSUE TITLE
Enforce Task dependencies

### ISSUE THREAD

**Issue by pankleks** [microsoft/vscode#32879]
I could not find a way to do it so I guess it's a feature request.

I have several tasks
1. build 1: tsc build project 1
2. build 2: tsc build project 2 (dependent on project 1)
3. styles: node-sass
4. clear: solution (deletes some files)

I wanted to create new task to do clean build of all. So I created new task using `dependsOn`
Problem is that new task is running all children in parallel with no dependencies which obviously does not work for me.

Proposal, change property `dependsOn` in following way:

```json
    "dependsOn": {
        "tasks": ["clear"],
        "then": {
            "tasks": [
                {
                    "tasks": ["build 1"],
                    "then": {
                        "tasks": ["build 2"]
                    }
                },
                {
                    "tasks": ["styles"]
                }
            ]
        }
    }
```
all tasks in `tasks` array at given level can be executed in parallel, after finishing go to `then` and proceed.

interface:

```ts
interface ITask {
    tasks: (ITask | string)[];
    then?: ITask;
}
```

Alternatively you could just add new property `subTask` - pointing to task that needs to be build before.
I would be enough to get dependencies, simpler to implement - but it would require to create a lot of fake tasks.

**Comment by dbaeumer** [comment]
This should work if:

Build 2 dependsOn Build 1
A generic tasks with only dependsOn (which is supported) depends on: build 2 and styles
clean dependsOn the generic task from above. 

Would that work for you?

**Comment by pankleks** [comment]
If I understand you right - you are suggesting to add `dependsOn` property to regular tasks (as tsc build for example) - right?

if so - yes - that would definitely work.

**Comment by dbaeumer** [comment]
Yes, that is the suggestion. The tsc tasks you define in the tasks.json will be 'merged' with the one provided by the TypeScript extension.

**Comment by pankleks** [comment]
Will that also work for non tsc tasks?

**Comment by dbaeumer** [comment]
Yes, it should also work for TSC tasks by 'customizing' the task in the tasks.json.

**Comment by pankleks** [comment]
Not sure what other information you need from me.

Your proposal is sound:) Please reopen.

**Comment by pankleks** [comment]
any plans for implementing it?

**Comment by dbaeumer** [comment]
This is currently not planned for any concrete milestone.

**Comment by pankleks** [comment]
ok ... how we can change it?:)

**Comment by dbaeumer** [comment]
Either by a pull request or by having more reactions on the item

**Comment by Spown** [comment]
how about a ``sync/async`` property that would determine how the ``dependsOn`` tasks are executed? I mean ``dependsOn`` is already an array so if we had something like this:
```json
        {
            "label": "rebuild",
            "dependsSync": true, 
            "dependsOn": ["foo", "bar"],
            "problemMatcher": []
        },
```
then ``bar`` task would be started only after ``foo`` is exited.

Further the road one could add another option that would start other tasks only if previous ones didn't fail, but for starters this could be at least something...

**Comment by Wyverex42** [comment]
I would also love to see this implemented. The solution proposed by Spown would be sufficient for me as well.

**Comment by alexr00** [comment]
I think the new-ish dependsOrder should do what you need: https://code.visualstudio.com/docs/editor/tasks#_compound-tasks
